core/src/main/java/org/elasticsearch/cockroach/CockroachActionCoordinator.java

```diff
@@ -0,0 +1,264 @@
+            listener);
+    }
+
+    /**
+     * Changes the executor node for a running cockroach task
+     *
+     * @param uuid         uuid of the task that should be updated
+     * @param executorNode the new executor node id
+     * @param listener     the listener that will be called when task is updated
+     */
+    public void updateCockroachTaskExecutor(String uuid, String executorNode, ActionListener<Empty> listener) {
+        performOperationOnCockroachTaskInProgress("update cockroach task executor", uuid,
+            taskInProgress -> new CockroachTaskInProgress<>(taskInProgress, executorNode, taskInProgress.getCallerNode()),
+            listener);
+    }
+
+    private <Request extends CockroachRequest<Request>, Response extends CockroachResponse> void performOperationOnCockroachTaskInProgress(
+        String source, String uuid,
+        Function<CockroachTaskInProgress<Request, Response>, CockroachTaskInProgress<Request, Response>> operation,
+        ActionListener<Empty> listener) {
+
+        clusterService.submitStateUpdateTask(source, new ClusterStateUpdateTask() {
+
+            @Override
+            public ClusterState execute(ClusterState currentState) throws Exception {
+                CockroachTasksInProgress cockroachTasksInProgress = currentState.custom(CockroachTasksInProgress.TYPE);
+                if (cockroachTasksInProgress == null) {
+                    logger.warn("Trying to update cockroach task with id {} that no longer exists, new master?", uuid);
+                    return currentState;
+                }
+                final List<CockroachTaskInProgress> currentTasks = new ArrayList<>();
+                boolean found = false;
+                for (CockroachTaskInProgress taskInProgress : cockroachTasksInProgress.entries()) {
+                    if (found == false && uuid.equals(taskInProgress.getUuid())) {
+                        found = true;
+                        //noinspection unchecked
+                        CockroachTaskInProgress newTask = operation.apply(taskInProgress);
+                        if (newTask != null) {
+                            currentTasks.add(newTask);
+                        }
+                    } else {
+                        currentTasks.add(taskInProgress);
+                    }
+                }
+                if (found) {
+                    ClusterState.Builder builder = ClusterState.builder(currentState);
+                    return builder.putCustom(CockroachTasksInProgress.TYPE, new CockroachTasksInProgress(currentTasks)).build();
+                } else {
+                    logger.warn("Trying to {} cockroach task with id {} that no longer exists", source, uuid);
+                    return currentState;
+                }
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                if (listener != null) {
```

Comment:
its not clear to me when `listener` could ever be null here (and also below in `clusterStateProcessed`)?  if it is possible, then the  method should have @Nullable annotation for listener